### PR TITLE
[#9793] Instructor/Courses/Details: Status in student list showing as "Unknown"

### LINKED ISSUE #9793: Instructor/Courses/Details: Status in student list showing as "Unknown"
- **Environment**: 
`master` branch at commit 30f56ffea3016decff6026cc46bb03e346ab7ffd

**Steps to reproduce**
1. Go to courses/details?courseid=ANY_VALID_COURSE_ID
2. Look at the entry under "Status" header in the student list

**Expected behaviour**
Should be either "Joined" or "Yet to Join"

**Actual behaviour**
Shows as Uknown

**Additional info**
Regression from PR #9777. 

Course details page should be updated to use the new REST endpoints.

### PR DESCRIPTION
**Pull request by erjillsison**
Fixes #9793

Part of #9595 

**Outline of Solution**
Update Instructor course details page to make use of new endpoints

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/output/InstructorData.java [hunk 1]
```diff
@@ -18,7 +18,7 @@
     public InstructorData(InstructorAttributes instructorAttributes) {
         this.courseId = instructorAttributes.getCourseId();
         this.email = instructorAttributes.getEmail();
-
+        this.role = instructorAttributes.getRole();
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/output/InstructorData.java]: I think you should hide the information to students.

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 2]
```diff
@@ -23,27 +33,18 @@ interface CourseStats {
 }
 
 interface CourseDetailsBundle {
-  course: CourseAttributes;
-  stats: CourseStats;
+  course?: CourseAttributes;
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: Is it necessary to add the optional mark `?`?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 3]
```diff
@@ -81,19 +82,106 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
...
+      this.statusMessageService.showErrorMessage(resp.error.message);
+    });
+  }
+
+  /**
+   * Loads the students in the course
+   */
+  private loadStudents(courseid: string): void {
+    this.studentService.getStudentsFromCourse(courseid).subscribe((students: Students) => {
+      const sections: StudentIndexedData = students.students.reduce((acc: StudentIndexedData, x: Student) => {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: There are better way to count the number of sections and teams? Use `Set()`?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 2 above)
Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: I suppose it's not necessary, just have to be more verbose i.e. initialize both `course` and `stats` to have empty properties. Would it be better that way?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 3 above)
Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: For sections, it's not only used to count the number of sections but to also group the students into sections. But for teams then yes, I'll update it to use `Set`.

By the way, `loadStudents` and `loadPrivilege` were borrowed from `instructor-student-list-page`

File: src/main/java/teammates/ui/webapi/output/InstructorData.java (hunk 1 above)
Comment (erjillsison) [reply src/main/java/teammates/ui/webapi/output/InstructorData.java]: Updated

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 2 above)
Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: It is not just for convenience. I consider CourseDetailsBundle as a model in the page. The model should have property `course` and `stats`. That is to say, if course exist, then there must be a stats associated with it. If you put ? before stats, I should read it as `stats for course is optional`

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 3 above)
Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: > For sections, it's not only used to count the number of sections but to also group the students into sections. 

You should remember long time ago, you are trying to flat the student list but fail :P So now you see how inconvenient it is.

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 2 above)
Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: That makes sense. Fixed

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 4]
```diff
@@ -7,9 +7,19 @@ import { CourseService } from '../../../services/course.service';
...
+  MessageOutput,
+  Student,
+  Students,
+} from '../../../types/api-output';
 import { ErrorMessageOutput } from '../../error-message-output';
-import { StudentListSectionData } from '../student-list/student-list-section-data';
+import { Intent } from '../../Intent';
+import { StudentListSectionData, StudentListStudentData } from '../student-list/student-list-section-data';
 
 interface CourseAttributes {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: I think you don't need the interface anymore?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 5]
```diff
@@ -27,23 +37,14 @@ interface CourseDetailsBundle {
   stats: CourseStats;
 }
 
-interface InstructorAttributes {
-  googleId: string;
+interface StudentIndexedData {
+  [key: string]: Student[];
+}
+
+interface InstructorData {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: Same for this?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts (hunk 4 above)
Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: Alright. I'll use the api-output types directly

File: src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap [hunk 6]
```diff
@@ -45,7 +46,7 @@ exports[`InstructorCourseDetailsPageComponent should snap with a course with one
               <p
                 class="form-control-static"
               >
-                CS101
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap]: Why does this line get disappeared? Forget to update the test case?

File: src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap [hunk 7]
```diff
@@ -64,7 +65,7 @@ exports[`InstructorCourseDetailsPageComponent should snap with a course with one
               <p
                 class="form-control-static"
               >
-                Introduction to CS
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap]: Same here?

File: src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap (hunk 6 above)
Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/__snapshots__/instructor-course-details-page.component.spec.ts.snap]: You're right. Updated the test case

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 8]
```diff
@@ -81,19 +84,94 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
   }
 
   /**
-   * Loads the course's details based on the given course ID and email.
+   * Loads the course's details based on the given course ID.
    */
   loadCourseDetails(courseid: string): void {
-    const paramMap: { [key: string]: string } = { courseid };
-    this.httpRequestService.get('/courses/details', paramMap).subscribe((resp: CourseInfo) => {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: And you can also remove the old endpoint?

Comment (erjillsison) [reply src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts]: Alright

**Pull request merged**